scripts/run_e2e_tests.py

```diff
@@ -222,397 +117,172 @@ def is_oppia_server_already_running():
-    python_portpicker, which is compatible with the portserver this
-    function starts, to request ports.
-
-    By "compatible" we mean that python_portpicker requests a port by
-    sending a request consisting of the PID of the requesting process
-    and expects a response consisting of the allocated port number. This
-    is the interface provided by this portserver.
-
-    Returns:
-        subprocess.Popen. The Popen subprocess object.
-    """
-    process = subprocess.Popen([
-        'python', '-m',
-        '.'.join(['scripts', 'run_portserver']),
-        '--portserver_unix_socket_address',
-        PORTSERVER_SOCKET_FILEPATH,
-    ])
-    return process
-
-
-def cleanup_portserver(portserver_process):
-    """Shut down the portserver.
-
-    We wait KILL_TIMEOUT_SECS seconds for the portserver to shut down
-    after sending CTRL-C (SIGINT). The portserver is configured to shut
-    down cleanly upon receiving this signal. If the server fails to shut
-    down, we kill the process.
-
-    Args:
-        portserver_process: subprocess.Popen. The Popen subprocess
-            object for the portserver.
-    """
-    _kill_process(portserver_process)
-
-
 def run_tests(args):
     """Run the scripts to start end-to-end tests."""
-    oppia_instance_is_already_running = is_oppia_server_already_running()
+    # TODO(#11549): Move this to top of the file.
+    import contextlib2
 
-    if oppia_instance_is_already_running:
+    if is_oppia_server_already_running():
         sys.exit(1)
-    setup_and_install_dependencies(args.skip_install)
 
-    common.start_redis_server()
-    atexit.register(cleanup)
+    install_third_party_libraries(args.skip_install)
 
-    dev_mode = not args.prod_env
+    with contextlib2.ExitStack() as stack:
+        dev_mode = not args.prod_env
 
-    if args.skip_build:
-        build.modify_constants(prod_env=args.prod_env)
-    else:
-        build_js_files(
-            dev_mode, deparallelize_terser=args.deparallelize_terser,
-            source_maps=args.source_maps)
-    version = args.chrome_driver_version or get_chrome_driver_version()
-    python_utils.PRINT('\n\nCHROMEDRIVER VERSION: %s\n\n' % version)
-    start_webdriver_manager(version)
+        if args.skip_build:
+            build.modify_constants(prod_env=args.prod_env)
+        else:
+            build_js_files(
+                dev_mode, deparallelize_terser=args.deparallelize_terser,
+                source_maps=args.source_maps)
+        stack.callback(build.set_constants_to_default)
 
-    # TODO(#11549): Move this to top of the file.
-    import contextlib2
-    managed_dev_appserver = common.managed_dev_appserver(
-        'app.yaml' if args.prod_env else 'app_dev.yaml',
-        port=GOOGLE_APP_ENGINE_PORT, log_level=args.server_log_level,
-        clear_datastore=True, skip_sdk_update_check=True,
-        env={'PORTSERVER_ADDRESS': PORTSERVER_SOCKET_FILEPATH})
+        python_utils.PRINT('Waiting for servers to come up...')
```

Comment:
I don't think we need this anymore now that the waiting happens in the individual process managers

Reply:
Done.